pywr/parameters/_parameters.pyx

```diff
@@ -306,50 +309,82 @@ cdef class TablesArrayParameter(IndexParameter):
+            elif node.shape[1] > self.scenario.size:
                 warnings.warn("The length of the second dimension of the tables Node is greater than the size of the specified Scenario. Not all data is being used!")
-        if shape[0] < len(self.model.timestepper):
+        if node.shape[0] < len(self.model.timestepper):
             raise IndexError("The length of the first dimension of the tables Node should be equal to or greater than the number of timesteps.")
 
+        # detect data type and read into memoryview
+        self._values_dbl = None
+        self._values_int = None
+        if self.scenario:
+            # if possible, only load the data required
+            scenario_indices = None
+            if self.model.scenarios.user_combinations:
+                scenario_indices = set()
+                for user_combination in self.model.scenarios.user_combinations:
+                    scenario_indices.add(user_combination[self._scenario_index])
+                scenario_indices = sorted(list(scenario_indices))
+                self._scenario_ids = np.ones(self.scenario.size, dtype=np.int32)
+                for n, i in enumerate(scenario_indices):
+                    self._scenario_ids[i] = n
+                if node.dtype in (np.float32, np.float64):
+                    self._values_dbl = node[:len(self.model.timestepper), scenario_indices]
+                else:
+                    self._values_int = node[:len(self.model.timestepper), scenario_indices]
+            elif self.scenario.slice:
+                self._scenario_ids = np.ones(self.scenario.size, dtype=np.int32)
+                scenario_indices = range(*self.scenario.slice.indices(self.scenario.slice.stop))
+                for n, i in enumerate(scenario_indices):
+                    self._scenario_ids[i] = n
+                if node.dtype in (np.float32, np.float64):
+                    self._values_dbl = node[:len(self.model.timestepper), self.scenario.slice]
+                else:
+                    self._values_int = node[:len(self.model.timestepper), self.scenario.slice]
+            # else: scenario is defined, but all data required
+        if node.dtype in (np.float32, np.float64):
+            if self._values_dbl is None:
+                self._values_dbl = node.read().astype(np.float64)
+            # negative values are often erroneous
+            if np.min(self._values_dbl) < 0.0:
+                warnings.warn('Negative values in input file "{}" from node: {}'.format(self.h5file, self.node))
+        else:
+            if self._values_int is None:
+                self._values_int = node.read().astype(np.int32)
+
     cpdef double value(self, Timestep ts, ScenarioIndex scenario_index) except? -1:
-        cdef int i = ts._index
-        cdef int j
+        cdef Py_ssize_t i = ts._index
+        cdef Py_ssize_t j
         if self._values_dbl is None:
             return float(self.get_index(scenario_index))
```

Comment:
It was like that where I got there! I think it's actually a bug. This is actually OK, because index is calculated before value. But the other (`get_value` call in `index`) will be broken.

https://github.com/pywr/pywr/blob/e74a0bd71964fdb8fdf214ef5e7d755100b985b1/pywr/parameters/_parameters.pyx#L595-L600

This should be fixed.